common/src/main/java/org/opensearch/sdk/PutDataObjectRequest.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.sdk;
+
+import org.opensearch.core.xcontent.ToXContentObject;
+
+public class PutDataObjectRequest {
+
+    private final String index;
+    private final ToXContentObject dataObject;
```

Comment:
Can java bounded generic types be used instead of XContentObject? 

https://docs.oracle.com/javase/tutorial/java/generics/bounded.html 

Generic type can enable strong type checking while performing CRUD operations.

Reply:
What value would that bring?

I did initially use generics, primarily to enable a typed return value from the GET implementation.  However, that failed because the `parse()` (or sometimes `fromXContent()`) implementations are static methods.  I could use reflection to call those methods, but they do not have a consistent signature.  

For example, `Connector` is an interface and has a static `createConnector()` method to do the parsing.   

I can't enforce signatures on static methods like I can instance methods with an interface... which I did try briefly with a `parse()` method required to be implemented (that just wrapped the static method).

But for me to call that I need to create an instance; but `Connector` is an interface so I'd have to create an anonymous implementation of all the abstract methods... which are very specific to `Connector` and impossible to put in a generic implementation.

So I finally just decided to return the `parser` from the Get method and let the caller use it to create the object.

That gave me a leftover `DataObject` that no longer used its parse method that you suggested removing [here](https://github.com/opensearch-project/ml-commons/pull/2430#discussion_r1602169911).

So what's the benefit here in an `Object<? extends ToXContentObject> foo` vs. `ToXContentObject foo`?

Reply:
Makes sense. Generic types would have helped with type checking but I agree that it not a primary requirement. Only additional challenge might be with client having to parse the request but opensearch plugins already operate this way, so it shouldn't be a concern.